lib/src/tag_detection.dart

```diff
@@ -78,27 +78,38 @@ import 'package:analyzer/dart/analysis/session.dart';
 import 'package:analyzer/dart/ast/ast.dart';
 import 'package:analyzer/dart/analysis/context_builder.dart';
 import 'package:analyzer/dart/analysis/context_locator.dart';
-import 'package:collection/collection.dart';
 import 'package:meta/meta.dart';
 import 'package:pana/pana.dart';
 import 'package:path/path.dart' as path;
 import 'package:pub_semver/pub_semver.dart';
 
-import 'logging.dart';
-
 abstract class _DirectedGraph<T> {
   Set<T> directSuccessors(T t);
 }
 
+/// Creates a suggestion indicating an issue, using [path] to give the location
+/// of the issue.
+typedef Explainer<N> = Suggestion Function(List<N> path);
+
+/// Returns an [Explainer] if node fullfills the predicate (if it violates some
```

Comment:
These should all be private - Done.

The idea is that the path is not ready at the point in the search where we find the 'bad' node (because we prepend old paths if we encounter a known bad node via `_PathFinder._cache`).